azavea/idb-osm-extraction-tool
# Download extracted OSM data as Shapefile

## Overview

This PR downloads the extracted OSM data from the stock response as a Shapefile in passing.

Currently it happens as part of the process of clicking the "Extract" button: clicking that kicks off a promise which now calls a `downloadShapefile` function before returning the points to plot on the map. While this one step extract may not be the final workflow, it seems like a good enough stopping point for now.

We use a forked version of the `shp-write` library from https://github.com/azavea/shp-write/tree/ki/add-ie-support -- which incorporates a patch to support downloading in IE11, too.

Connects #8

### Demo

<img width="811" alt="screen shot 2018-07-18 at 4 29 19 pm" src="https://user-images.githubusercontent.com/4165523/42906016-c3ad9564-8aa7-11e8-8ea9-e8bd9775aef3.png">

<img width="670" alt="screen shot 2018-07-18 at 4 21 10 pm" src="https://user-images.githubusercontent.com/4165523/42905988-aec86c64-8aa7-11e8-980f-c324b377d6cd.png">

![screen shot 2018-07-18 at 4 23 04 pm](https://user-images.githubusercontent.com/4165523/42905996-b221c7d4-8aa7-11e8-9407-5ce79aed211b.png)

### Notes

As mentioned above and also here -- https://github.com/azavea/idb-osm-extraction-tool/pull/18#pullrequestreview-138022414 -- there's a potential alternate workflow for this which would be:

- click a button retrieve data and plot on the map
- click another button to extract data into a shapefile

I think that seems reasonable but I'd like to defer thinking about it til we regroup on the remainder of the tasks here.

## Testing Instructions

- get this branch, then vaporize your node_modules dir
- `./scripts/update` and then `./scripts/server` and in both Chrome and IE11...
   - select an AOI including Georgetown
   - click "Extract" and verify that the completed result both plots the circle markers on the map and also downloads a zip file, currently called "buildings.zip" (but in future likely to be feature + date range)
   - open the downloaded zip files and verify that they include the Shapefile data
   - open one of the Shapefiles in QGIS with an OSM basemap and verify that the data is there
